tools/cargo-each/src/config.rs

```diff
@@ -0,0 +1,173 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use cargo_metadata::Package;
+use serde::{Deserialize, Deserializer};
+
+#[derive(Debug)]
+pub(crate) struct Config<'r> {
+    pub combinations: Vec<Combination<'r>>,
+    pub sets: BTreeMap<&'r str, Set<'r>>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "kebab-case", untagged, deny_unknown_fields)]
+pub(crate) enum IdentOrList<'r> {
+    Ident(#[serde(borrow)] &'r str),
+    List(#[serde(borrow, deserialize_with = "deserialize_set")] BTreeSet<&'r str>),
+}
+
+#[derive(Debug)]
+pub(crate) struct Combination<'r> {
+    pub groups: BTreeSet<&'r str>,
+    pub sets: BTreeSet<&'r str>,
+    pub always_on: BTreeSet<&'r str>,
+    pub include_rest: bool,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+pub(crate) struct Set<'r> {
+    #[serde(borrow, deserialize_with = "deserialize_set", default)]
+    pub members: BTreeSet<&'r str>,
+    pub at_least_one: bool,
+    pub mutually_exclusive: bool,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+struct GenericConfig<'r> {
+    #[serde(borrow)]
+    combinations: Vec<GenericCombination<'r>>,
+    #[serde(borrow, default)]
+    sets: BTreeMap<&'r str, Set<'r>>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+struct GenericCombination<'r> {
+    #[serde(borrow, default)]
+    generics: BTreeMap<&'r str, IdentOrList<'r>>,
+    #[serde(borrow, deserialize_with = "deserialize_set", default)]
+    groups: BTreeSet<&'r str>,
+    #[serde(borrow, deserialize_with = "deserialize_set", default)]
+    sets: BTreeSet<&'r str>,
+    #[serde(borrow, deserialize_with = "deserialize_set", default)]
+    always_on: BTreeSet<&'r str>,
+    include_rest: bool,
+}
+
+fn deserialize_set<'r, 'de: 'r, D>(deserializer: D) -> Result<BTreeSet<&'r str>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    Vec::<&'r str>::deserialize(deserializer)
+        .map(Vec::into_iter)
+        .map(BTreeSet::from_iter)
+}
+
+pub(crate) fn deserialize_config_if_any(package: &Package) -> Result<Option<Config<'_>>, ()> {
+    package
+        .metadata
+        .get("cargo-each")
+        .map(GenericConfig::deserialize)
+        .transpose()
+        .map_err(
+            #[cold]
+            |error| eprintln!("Deserializing configuration failed!\nError: {error}"),
+        )
+        .and_then(|maybe_config| {
+            maybe_config
+                .map(|config| try_resolve_config_generics(package, config))
+                .transpose()
+        })
+}
+
+fn try_resolve_config_generics<'r>(
```

Comment:
~90 LoC for a function is too much

Reply:
I have forgotten to break it down. My bad.